external-crates/move/crates/move-compiler/src/naming/translate.rs

```diff
@@ -456,43 +464,101 @@ impl<'env> Context<'env> {
+        &mut self,
+        loc: Loc,
+        name: Option<P::BlockLabel>,
+        name_type: NominalBlockType,
+    ) {
         debug_assert!(
             self.nominal_blocks.len() < 100,
             "Nominal block list exceeded 100."
         );
-        let sym = if let Some(name) = name {
-            name.value()
-        } else {
-            // all named blocks have names, so a non-named block must be a loop
-            N::Exp_::LOOP_NAME_SYMBOL
-        };
         let id = self.nominal_block_id;
         self.nominal_block_id += 1;
-        self.nominal_blocks.push((sym, id, name_type));
+        self.nominal_blocks
+            .push((loc, name.map(|n| n.value()), id, name_type));
+    }
+
+    fn current_loop(
+        &self,
+        loc: Loc,
+        usage: NominalBlockUsage,
+    ) -> Result<BlockLabel, Option<Diagnostic>> {
+        let Some((lblloc, name, id, name_type)) =
+            self.nominal_blocks
+                .iter()
+                .rev()
+                .find(|(_, _, _, name_type)| {
+                    matches!(
+                        name_type,
+                        NominalBlockType::Loop(_) | NominalBlockType::Lambda
+                    )
+                })
+        else {
+            return Err(None);
+        };
+        if *name_type == NominalBlockType::Lambda {
+            // lambdas capture break/continue even though it is not yet supported
+            let msg =
+                format!("Invalid '{usage}'. This usage is not yet supported for lambdas or macros");
+            let mut diag = diag!(
+                TypeSafety::InvalidLoopControl,
+                (loc, msg),
+                (*lblloc, "Inside this lambda")
+            );
+            // suggest adding a label to the loop
+            let most_recent_loop_opt =
+                self.nominal_blocks
+                    .iter()
+                    .rev()
+                    .find_map(|(loc, name, _, name_type)| {
+                        if let NominalBlockType::Loop(loop_type) = name_type {
+                            Some((*loc, name, *loop_type))
+                        } else {
+                            None
+                        }
+                    });
+            if let Some((loop_loc, name, loop_type)) = most_recent_loop_opt {
+                let msg = if let Some(loop_label) = name {
+                    format!(
+                        "To '{usage}' to this loop, specify the label, \
+                        e.g. `{usage}'{loop_label}`",
+                    )
+                } else {
+                    format!(
+                        "To '{usage}' to this loop, add a label, \
+                        e.g. `'label: {loop_type}` and `{usage}'label`",
+                    )
+                };
+                diag.add_secondary_label((loop_loc, msg));
+            }
+            return Err(Some(diag));
+        }
+        Ok(block_label(loc, usage, *name, *id, *name_type))
+    }
```

Comment:
The result here I think is rather strange. I would prefer to move the diagnostic work from the traversal below to here, to make things nicely uniform. I did it in the traversal originally to avoid passing the usage type, but since we've already done it we might as well embrace it IMO.

```suggestion
    ) -> Option<BlockLabel> {
        let label_info =
            self.nominal_blocks
                .iter()
                .rev()
                .find(|(_, _, _, name_type)| {
                    matches!(
                        name_type,
                        NominalBlockType::Loop(_) | NominalBlockType::LambdaBreak
                    )
                });
        if let Some((name, label, name_type)) = label_info {
            match name_type {
                NominbalBlockType::Loop(_) => name,
                NominalBlockType::LambdaBreak => {
                    // lambdas capture break/continue even though it is not yet supported
                    let msg =
                        format!("Invalid '{usage}'. This usage is not yet supported for lambdas or macros");
                    let mut diag = diag!(
                        TypeSafety::InvalidLoopControl,
                        (loc, msg),
                        (label.var.loc, "Inside this lambda")
                    );
                    // suggest adding a label to the loop
                    let most_recent_loop_opt =
                        self.nominal_blocks
                            .iter()
                            .rev()
                            .find_map(|(loc, name, _, name_type)| {
                                if let NominalBlockType::Loop(loop_type) = name_type {
                                    Some((*loc, name, *loop_type))
                                } else {
                                    None
                                }
                            });
                    if let Some((loop_loc, name, loop_type)) = most_recent_loop_opt {
                        let msg = if let Some(loop_label) = name {
                            format!(
                                "To '{usage}' to this loop, specify the label, \
                                e.g. `{usage}'{loop_label}`",
                            )
                        } else {
                            format!(
                                "To '{usage}' to this loop, add a label, \
                                e.g. `'label: {loop_type}` and `{usage}'label`",
                            )
                        };
                        diag.add_secondary_label((loop_loc, msg));
                    }
                    self.env.add_diag(diag);
                    None
                }
                _ => panic!("ICE ... msg ... "),
            }
        } else {
            match usage {
                NominalBlockUsage::Continue => {
                        let diag = {
                            let msg = "Invalid usage of 'continue'. \
                            'continue' can only be used inside a loop body or lambda";
                            diag!(TypeSafety::InvalidLoopControl, (loc, msg))
                        };
                        context.env.add_diag(diag);
                },
                NominalBlockUsage::Break => {
                        let diag = {
                            let msg = "Invalid usage of 'break'. \
                            'break' can only be used inside a loop body or lambda";
                            diag!(TypeSafety::InvalidLoopControl, (loc, msg))
                        };
                        self.env.add_diag(diag);
                },
                _ => panic!("ICE ... msg ... "),
            }
            None
        }
    }
```